hieroplatform/src/main/java/org/hiero/sketch/spreadsheet/HeatMap.java

```diff
@@ -0,0 +1,132 @@
+package org.hiero.sketch.spreadsheet;
+
+import org.checkerframework.checker.nullness.qual.NonNull;
+import org.hiero.sketch.table.api.IColumn;
+import org.hiero.sketch.table.api.IMembershipSet;
+import org.hiero.sketch.table.api.IRowIterator;
+import org.hiero.sketch.table.api.IStringConverter;
+import org.hiero.utils.IntSet;
+
+/**
+ * An implementation of a 2 dimension histogram. It is designed assuming the number of buckets is very large, so there
+ * are no bucket objects, nor min and max items per bucket, the only thing stored is counts.
+ * The buckets are assumed to be equisize.
+ */
+public class HeatMap {
+    private final long[][] buckets;
+    private long missingData; //number of items missing on both columns
+    private long outOfRange;
+    private final IBucketsDescription1D bucketDescDim1;
+    private final IBucketsDescription1D bucketDescDim2;
+    private Histogram1DLight histogramMissingD1; // hist of items that are missing in D2
+    private Histogram1DLight histogramMissingD2; // hist of items that are missing in D1
+    private boolean initialized;
+    private long totalsize;
+
+    public HeatMap(final @NonNull IBucketsDescription1D buckets1,
+                   final @NonNull IBucketsDescription1D buckets2) {
+        this.bucketDescDim1 = buckets1;
+        this.bucketDescDim2 = buckets2;
+        this.buckets = new long[buckets1.getNumOfBuckets()][buckets2.getNumOfBuckets()]; // Automatically initialized to 0
+        this.initialized = false;
+        histogramMissingD1 = new Histogram1DLight(bucketDescDim1);
+        histogramMissingD2 = new Histogram1DLight(bucketDescDim2);
+    }
+
+    /**
+     * Creates the histogram explicitly and in full. Should be called at most once.
+     */
+    public void createHistogram(final IColumn columnD1, final IColumn columnD2,
+                                final IStringConverter converterD1, final IStringConverter converterD2,
+                                final IMembershipSet membershipSet) {
+        if (this.initialized) //a histogram had already been created
+            throw new IllegalAccessError("A histogram cannot be created twice");
+        this.initialized = true;
+        final IRowIterator myIter = membershipSet.getIterator();
+        int currRow = myIter.getNextRow();
+        while (currRow >= 0) {
+            if ((columnD1.isMissing(currRow)) || (columnD2.isMissing(currRow))){
+                if (!columnD1.isMissing(currRow))  //only column 2 is missing
```

Comment:
Maybe you can call isMissing only once for each column and save the result in a temporary bool.